### PR TITLE
Remove the JS stubs for main module imports by creating them dynamically

### PR DESCRIPTION
**Pull request by awtcode**
The goal of this PR is to remove the JS stubs that are hard-coded into the Javascript loader because large applications can potentially generate a large amount of these stubs.

The idea here is to generate them dynamically using a proxy handler for the main module. However, there is a bug now whereby minified imports and exports between Javascript and Wasm are not honoured because the names of missing functions are removed from the Javascript loader.

@kripken and @sbc100 , this is not ready for submission yet but I thought it will be good to create a PR just to discuss the approach first. Thanks.

### CODE REVIEW

File: src/preamble.js [hunk 1]
```diff
@@ -978,12 +978,38 @@ function createWasm(env) {
...
+#endif
+      return Module[prop].apply(null, arguments);
+    };
+  }
+};
 
   // prepare imports
   var info = {
-    'env': env
+    env: new Proxy(env, proxyHandler),
```
Comment (sbc100) [inline src/preamble.js]: Don't you need to quote the 'env' here?

File: emscripten.py [hunk 2]
```diff
@@ -97,7 +97,7 @@ def emscript_fastcomp(infile, outfile, memfile, libraries, compiler_engine,
 
     with ToolchainProfiler.profile_block('function_tables_and_exports'):
       (post, function_table_data, bundled_args) = (
-          function_tables_and_exports(funcs, metadata, mem_init, glue, forwarded_data, outfile, DEBUG))
+          function_tables_and_exports(funcs, metadata, mem_init, glue, forwarded_data, outfile, DEBUG))
```
Comment (sbc100) [inline emscripten.py]: This line looks like it can be reverted?

File: src/preamble.js [hunk 3]
```diff
@@ -978,12 +978,38 @@ function createWasm(env) {
...
+      // and the dynamic library cannot provide JS for it. Use
+      // the generic "X" invoke for it.
+      return obj[prop] = invoke_X;
+    }
+
+    if (Module[prop]) {
+      return Module[prop];
+    }
+
+    // if not a global, then a function - call it indirectly
```
Comment (sbc100) [inline src/preamble.js]: This comment seems wrong.

Seems like it should say something like:  "For a missing property generate a stub that will do a runtime lookup and error out if its still missing".

File: src/jsifier.js [hunk 4]
```diff
@@ -1,549 +1,561 @@
...
+          }
+
+          var target = (SIDE_MODULE ? 'parent' : '') + "Module['_" + realIdent + "']";
+          var assertion = '';
+          if (ASSERTIONS) {
+            var what = 'function';
+            if (isGlobalAccessor) {
+              what = 'global';
+            }
+            assertion += 'if (!' + target + ') abort("external ' + what + ' \'' + realIdent + '\' is missing. perhaps a side module was not linked in? if this function was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment");\n';
```
Comment (sbc100) [inline src/jsifier.js]: Something seems to have gone wrong with the diff here.. maybe you changed the line endings to windows by accident?

File: src/jsifier.js [hunk 5]
```diff
@@ -1,549 +1,561 @@
...
+          });
+          proxiedFunctionTable.push(finalName);
+        } else {
+          contentText = snippet; // Regular JS function that will be executed in the context of the calling thread.
+        }
+      } else if (typeof snippet === 'string' && snippet.indexOf(';') == 0) {
+        contentText = 'var ' + finalName + snippet;
+        if (snippet[snippet.length-1] != ';' && snippet[snippet.length-1] != '}') contentText += ';';
+      } else if (typeof snippet === 'undefined') {
+        contentText = '';
```
Comment (sbc100) [inline src/jsifier.js]: Why is this change needed?

File: emscripten.py (hunk 2 above)
Comment (awtcode) [reply emscripten.py]: Yes, I will revert the extra space here.

File: src/preamble.js (hunk 1 above)
Comment (awtcode) [reply src/preamble.js]: Thanks, I will quote it although it doesn't make a difference functionally.

File: src/jsifier.js (hunk 4 above)
Comment (awtcode) [reply src/jsifier.js]: I think its because I forgot to do another merge from incoming before sending the review. Does it look ok now?

File: src/jsifier.js (hunk 5 above)
Comment (awtcode) [reply src/jsifier.js]: Without this, the code will go down to the block below and we end up with
`var sidey=undefined;`

**Comment by awtcode** [comment]
> I'm not sure I understand the purpose of the new sideFunctions mapping, but it looks like its going in the right direction.

I hope this is clear after today's meeting. The purpose is to mark the missing function so that we won't generate the stub for it and pass it to asmLibraryArg.

Perhaps `missingFunctions` would be more intuitive?

**Comment by awtcode** [comment]
@sbc100 , I've updated the PR to fix up the jsifier and to handle the minification step for the imports/exports. Let me know if you have any thoughts on this. Thanks.

File: src/preamble.js (hunk 3 above)
Comment (awtcode) [reply src/preamble.js]: Sure, will update it.

File: src/jsifier.js [hunk 6]
```diff
@@ -196,11 +196,18 @@ function JSify(data, functionsOnly) {
...
-          LibraryManager.library[ident] = new Function(functionBody);
+	        
           if (SIDE_MODULE) {
+            LibraryManager.library[ident] = new Function(functionBody);
             // no dependencies, just emit the thunk
             Functions.libraryFunctions[finalName] = 1;
             return processLibraryFunction(LibraryManager.library[ident], ident, finalName);
+          } else {
+            // The JS stubs are still generated for asm.js because we are not loading the asm.js code dynamically
+            if (WASM)
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/jsifier.js]: Sure, will add it in.

File: src/preamble.js [hunk 7]
```diff
@@ -978,12 +978,33 @@ function createWasm(env) {
...
+#endif
+      return Module[funcName].apply(null, arguments);
+    };
+  }
+};
 
   // prepare imports
   var info = {
-    'env': env
+    'env': new Proxy(env, proxyHandler),
```
[the remark this answers is not in the document]
Comment (awtcode) [reply src/preamble.js]: Sure @kripken , I can do this. But this is a tradeoff between what we currently have e.g.
`var __ZN35ClassAManagerD0Ev= function() {
  Module["__ZN35ClassAManagerD0Ev"].apply(null, arguments);
}

var __ZN35ClassBManagerD0Ev= function() {
  Module["__ZN35ClassBManagerD0Ev"].apply(null, arguments);
}
.
.
.
var asmLibraryArg = {
  "abort": abort,
  "bMf": __ZN35ClassAManagerD0Ev,
  "Qhi": __ZN35ClassBManagerD0Ev,
`
and

`var proxyHandler = {
  'get': function(obj, prop) {
    if (prop in obj) {
      return obj[prop]; // already present
    }

    // Check if the minification mapping exists first, if not just
    // use the original name.
    var funcName = prop;
    if (Module["mapping"] && Module["mapping"][prop]) {
      funcName = prop;
    }
    
    // For a missing property generate a stub that will do a runtime lookup and error out if its still missing
    return env[prop] = function() {
      return Module[funcName].apply(null, arguments);
    };
  }
};
.
.
.
Module.mapping={bMf:"__ZN35ClassAManagerD0Ev",Qhi:"__ZN35ClassBManagerD0Ev"...};`

Unless we are looking at applications with a very small number of imports, most of the time, I believe the current approach with the JS stubs will end up with a larger JS file size? But I'm fine with putting this behind a flag as well if this is not the case. Just let me know. Thanks.

File: tools/shared.py [hunk 8]
```diff
@@ -2561,6 +2575,7 @@ def minify_wasm_imports_and_exports(js_file, wasm_file, minify_whitespace, minif
     if minify_whitespace:
       passes.append('minifyWhitespace')
     extra_info = {'mapping': mapping}
+    Building.write_mapping_to_file(js_file, mapping)
```
[the remark this answers is not in the document]
Comment (awtcode) [reply tools/shared.py]: My comments are as above.

File: src/preamble.js [hunk 9]
```diff
@@ -978,12 +978,33 @@ function createWasm(env) {
...
+var proxyHandler = {
+  'get': function(obj, prop) {
+    if (prop in obj) {
+      return obj[prop]; // already present
+    }
+
+    // Check if the minification mapping exists first, if not just
+    // use the original name.
+    var funcName = prop;
+    if (Module["mapping"] && Module["mapping"][prop]) {
```
Comment (kripken) [inline src/preamble.js]: btw, it would be better to not put this on `Module`. It can be just `var envMapping` at the global scope. That would allow better minification.

(Unless this does need to be exported to the outside for some reason?)

File: src/preamble.js [hunk 10]
```diff
@@ -978,12 +978,33 @@ function createWasm(env) {
...
+      Module["mapping"][prop];
+    }
+    
+    // For a missing property generate a stub that will do a runtime lookup and error out if its still missing
+    return env[prop] = function() {
+#if ASSERTIONS
+      assert(Module[funcName], 'missing linked function ' + funcName + '. perhaps a side module was not linked in? if this function was expected to arrive from a system library, try to build the MAIN_MODULE with EMCC_FORCE_STDLIBS=1 in the environment');
+#endif
+      return Module[funcName].apply(null, arguments);
+    };
```
Comment (kripken) [inline src/preamble.js]: It looks like this change turns startup errors into lazy errors later on when actually called - that is only needed with dynamic linking, correct? So we can put it behind `MAIN_MODULE` I think.

File: src/preamble.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (kripken) [reply src/preamble.js]: Thanks, now I see what you mean. Yes, this is an interesting tradeoff.

Do you have some numbers for code size with your app, and with hello world and maybe some other things (at `-O3`)? Would be good to get an idea of what kind of size differences are here.

**Comment by awtcode** [comment]
@kripken and @sbc100, I'm currently trying to stabilize the tests but I see multiple 'process out of memory' errors that I cannot repro on my Windows machine:

https://circleci.com/gh/emscripten-core/emscripten/176104?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

https://circleci.com/gh/emscripten-core/emscripten/176102?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

Are these intermittent errors and do you have any suggestions on how to resolve them? Thanks.

File: src/preamble.js (hunk 10 above)
Comment (awtcode) [reply src/preamble.js]: Sure, will add this in.

**Comment by kripken** [comment]
Hmm, another idea on top of the one I wrote in the email (about the node version) - we updated node recently. If latest incoming hasn't been merged in to here, that might help.

**Comment by awtcode** [comment]
> Hmm, another idea on top of the one I wrote in the email (about the node version) - we updated node recently. If latest incoming hasn't been merged in to here, that might help.

Sorry @kripken , after some more debugging. I realized that this issue can be reproduced on Windows Node v12.10.0 as well as Linux Node v12.9.1. This is not reproducible on Windows Node v10.16.3. I have already logged an issue against Node over here: https://github.com/nodejs/node/issues/29533

Would it make sense to downgrade Node back to v10?

**Comment by kripken** [comment]
I don't think we should downgrade - @sbc100 worked hard to upgrade it recently.

We may need to disable the test in v8/node.

However, it's odd that this happens - do you know why this test ends up hitting that bug? The code change here should be pretty minor? Maybe something's going wrong and we're emitting more code / GC stuff.

**Comment by awtcode** [comment]
> I don't think we should downgrade - @sbc100 worked hard to upgrade it recently.
> 
> We may need to disable the test in v8/node.
> 
> However, it's odd that this happens - do you know why this test ends up hitting that bug? The code change here should be pretty minor? Maybe something's going wrong and we're emitting more code / GC stuff.

I believe its because I'm generating the stubs dynamically in my PR and currently, we create a new Wasm module to convert each JS function to a Wasm export. In Node v12, we meet with the process out of memory error when this happens.

**Comment by bnoordhuis** [comment]
To confirm: that's what happens and with newer V8 versions each `WebAssembly.Module` reserves 1 GB of address space on 64 bits architectures. This PR creates so many that it exhausts the address space.

@awtcode opened https://github.com/nodejs/node/issues/29533 but it's really an upstream V8 issue and I suspect they don't consider it a bug but a performance feature - the reservation allows direct jumps in the generated machine code.

I would suggest using a single `WebAssembly.Module` if that's possible.

**Comment by kripken** [comment]
Ah, thanks @awtcode! Where in this PR do we start to create a new module for each JS function? I think I misunderstood that part. Thanks @bnoordhuis, yes, we should try to optimize that part.

I think [this is that issue](https://bugs.chromium.org/p/v8/issues/detail?id=9678), and if so then there is progress on fixing it. But given that it's still open, I don't think we can land a PR that hits this bug - we'd bother a lot of users. So hopefully we can optimize things.

File: src/preamble.js [hunk 11]
```diff
@@ -1002,12 +987,89 @@ function createWasm(env) {
...
+        // the fp$ function returns the address (table index) of the function
+        var parts = prop.split('$');
+        assert(parts.length == 3)
+        var name = parts[1];
+        var sig = parts[2];
+        var fp = 0;
+        return obj[prop] = function() {
+          if (!fp) {
+            var f = resolveSymbol(name, 'function');
+            fp = addFunctionWasm(f, sig);
```
Comment (awtcode) [inline src/preamble.js]: @kripken , this is where we dynamically generate the stubs for the functions pointed to by the fp$ functions and we subsequently add the function to the Wasm module which results in the OOM crash.

File: emscripten.py [hunk 12]
```diff
@@ -1781,7 +1768,7 @@ def create_fp_accessors(metadata):
   # For SIDE_MODULEs these are generated by the proxyHandler at runtime.
   accessors = []
   for fullname in metadata['declares']:
-    if not fullname.startswith('fp$'):
+    if not fullname.startswith('fp$') or [key for (key, value) in shared.Building.missing_imports.items() if fullname in key]:
```
Comment (sbc100) [inline emscripten.py]: I think you can write this as

```
any(fullname in key for key in shared.Building.missing_imports)
```

File: src/preamble.js (hunk 11 above)
Comment (sbc100) [reply src/preamble.js]: But this change should not have increased the number of $fp calls.. the only change should have been that before the addFunctionWasm calls were all static, and they are dynamic.

Perhaps that problem here related to JS scoping rules?  I think perhaps `var fp` is not scoped that way I thought it was.  The idea here was to create new fp for each function, but I think I was maybe wrong about how JS works :(

Comment (sbc100) [reply src/preamble.js]: But addFunctionWasm should always be called exactly once for each $fp wasm import.. both before and after this change IIUC

File: emscripten.py [hunk 13]
```diff
@@ -1797,7 +1784,7 @@ def create_fp_accessors(metadata):
   var func = Module['%(mangled)s'];
   if (!func)
     func = %(mangled)s;
-  var fp = addFunction(func, '%(sig)s');
+  var fp = addFunctionWasm(func, '%(sig)s');
```
Comment (kripken) [inline emscripten.py]: There are two changes like this in this PR. Perhaps they have an effect? In fastcomp, `addFunction` will not call `addFunctionWasm` necessarily, it depends on `EMULATED_FUNCTION_POINTERS`.

If that's not it, then one way to debug the OOM issue is to add some logging of each call to `addFunctionWasm`, and to view that logging before and after this PR.

**Comment by sbc100** [comment]
The deletion of the `incoming` branch triggered the automatic closing of this branch.  My apologies.  This was an unexpected side effect.

Re-opening and re-targeting to master.

**Comment by sbc100** [comment]
I think I achieved at least part of this by enabling the DECLARE_ASM_MODULE_EXPORTS=0 option for MAIN_MODULEs.   See #10348.  Can you try this and see if it does what you need?

If not can you rebase and re-upload?
